File: cli/src/freepod/table.py
"""Shared rendering for the client's listing commands.

A leaf module: `builds` and `releases` both draw the same kind of table, and
neither should have to import the other to do it.
"""

from __future__ import annotations

from datetime import datetime, timedelta, timezone
from typing import Any, List, Optional, Sequence

#: Digest characters kept when abbreviating an image reference. Twelve is the
#: conventional short form and is what makes a column of them comparable at a
#: glance; `--verbose` prints the reference in full.
SHORT_DIGEST = 12

#: Column separator. Two spaces, so a value can never be mistaken for a
#: boundary and `awk` still splits the rows.
GAP = "  "

#: Shown where the platform has no value yet — a queued build has no duration,
#: a failed one has no image.
BLANK = "-"


def parse_time(value: Any) -> Optional[datetime]:
    """One of the platform's timestamps, as an aware UTC datetime.

    The API serializes naive datetimes that are UTC by construction, so a
    missing offset is read as UTC rather than as local time — reading it as
    local would misreport every duration by the reader's own offset.
    """
    if not isinstance(value, str) or not value:
        return None
    text = value[:-1] + "+00:00" if value.endswith("Z") else value
    try:
        stamp = datetime.fromisoformat(text)
    except ValueError:
        return None
    return stamp.replace(tzinfo=timezone.utc) if stamp.tzinfo is None else stamp


def format_time(value: Any) -> str:
    """A timestamp in the reader's own timezone, to the minute."""
    stamp = parse_time(value)
    if stamp is None:
        return str(value) if value else BLANK
    return stamp.astimezone().strftime("%Y-%m-%d %H:%M")


def format_duration(delta: Optional[timedelta]) -> str:
    """`45s`, `3m 12s`, `1h 4m` — two units at most, largest first."""
    if delta is None:
        return BLANK
    seconds = int(delta.total_seconds())
    if seconds < 0:
        return BLANK
    if seconds < 60:
        return f"{seconds}s"
    minutes, seconds = divmod(seconds, 60)
    if minutes < 60:
        return f"{minutes}m {seconds}s"
    hours, minutes = divmod(minutes, 60)
    return f"{hours}h {minutes}m"


def elapsed(
    started: Any, finished: Any, now: Optional[datetime] = None
) -> Optional[timedelta]:
    """How long the work ran, or has been running.

    Measured from when work *started*, never from when the record was created:
    the wait for a worker is queueing, and counting it would report a
    five-second job as a five-minute one whenever the queue was busy.
    """
    begin = parse_time(started)
    if begin is None:
        return None
    end = parse_time(finished)
    if end is None:
        return (now or datetime.now(timezone.utc)) - begin
    return end - begin


def abbreviate(image: Any, *, full: bool = False) -> str:
    """Shorten a digest reference to its first twelve characters.

    A full reference is 75 characters of which 64 are a digest, which would
    make the column wider than every other one put together. Truncation is
    marked, never silent.
    """
    if not isinstance(image, str) or not image:
        return BLANK
    if full:
        return image
    head, marker, digest = image.rpartition("sha256:")
    if marker and len(digest) > SHORT_DIGEST:
        return f"{head}{marker}{digest[:SHORT_DIGEST]}…"
    return image


def render(table: Sequence[Sequence[str]]) -> str:
    """Left-aligned columns, sized to their contents, with no trailing space."""
    if not table:
        return ""
    columns = len(table[0])
    widths = [max(len(row[column]) for row in table) for column in range(columns)]
    lines: List[str] = [
        GAP.join(cell.ljust(width) for cell, width in zip(row, widths)).rstrip()
        for row in table
    ]
    return "\n".join(lines)

File: cli-rust/src/table.rs
//! Shared rendering for the client's listing commands.
//!
//! A leaf module: `history` and `releases` both draw the same kind of table.
//! Mirrors `table.py`.

use chrono::{DateTime, Duration, TimeZone, Utc};
use serde_json::Value;

/// Digest characters kept when abbreviating an image reference.
pub const SHORT_DIGEST: usize = 12;

/// Column separator. Two spaces.
pub const GAP: &str = "  ";

/// Shown where the platform has no value yet.
pub const BLANK: &str = "-";

/// A JSON value the way Python's `str()` reads it: a string bare, a number as
/// its digits, and a missing/null value as the blank.
pub fn value_str(v: &Value) -> String {
    match v {
        Value::String(s) => s.clone(),
        Value::Number(n) => n.to_string(),
        Value::Bool(b) => b.to_string(),
        Value::Null => BLANK.to_string(),
        other => other.to_string(),
    }
}

/// One of the platform's timestamps, as an aware UTC datetime.
///
/// The API serializes naive datetimes that are UTC by construction, so a
/// missing offset is read as UTC rather than as local time.
pub fn parse_time(value: &Value) -> Option<DateTime<Utc>> {
    let text = value.as_str()?;
    if text.is_empty() {
        return None;
    }
    let text = if let Some(stripped) = text.strip_suffix('Z') {
        format!("{stripped}+00:00")
    } else {
        text.to_string()
    };
    match DateTime::parse_from_rfc3339(&text) {
        Ok(dt) => Some(dt.with_timezone(&Utc)),
        Err(_) => {
            // Naive: read as UTC.
            if let Ok(naive) = chrono::NaiveDateTime::parse_from_str(&text, "%Y-%m-%dT%H:%M:%S%.f") {
                Some(Utc.from_utc_datetime(&naive))
            } else {
                None
            }
        }
    }
}

/// A timestamp in the reader's own timezone, to the minute.
pub fn format_time(value: &Value) -> String {
    match parse_time(value) {
        None => match value.as_str() {
            Some(s) if !s.is_empty() => s.to_string(),
            _ => BLANK.to_string(),
        },
        Some(stamp) => {
            let local = chrono::Local.from_utc_datetime(&stamp.naive_utc());
            local.format("%Y-%m-%d %H:%M").to_string()
        }
    }
}

/// `45s`, `3m 12s`, `1h 4m` — two units at most, largest first.
pub fn format_duration(delta: Option<Duration>) -> String {
    let Some(delta) = delta else {
        return BLANK.to_string();
    };
    let seconds = delta.num_seconds();
    if seconds < 0 || seconds < 60 {
        if seconds < 0 {
            return BLANK.to_string();
        }
        return format!("{seconds}s");
    }
    let minutes = seconds / 60;
    let seconds = seconds % 60;
    if minutes < 60 {
        return format!("{minutes}m {seconds}s");
    }
    let hours = minutes / 60;
    let minutes = minutes % 60;
    format!("{hours}h {minutes}m")
}

/// How long the work ran, or has been running.
///
/// Measured from when work *started*, never from when the record was created.
pub fn elapsed(
    started: &Value,
    finished: &Value,
    now: Option<DateTime<Utc>>,
) -> Option<Duration> {
    let begin = parse_time(started)?;
    let end = parse_time(finished);
    match end {
        None => Some((now.unwrap_or_else(Utc::now)) - begin),
        Some(e) => Some(e - begin),
    }
}

/// Shorten a digest reference to its first twelve characters.
pub fn abbreviate(image: &Value, full: bool) -> String {
    let Some(image) = image.as_str() else {
        return BLANK.to_string();
    };
    if image.is_empty() {
        return BLANK.to_string();
    }
    if full {
        return image.to_string();
    }
    match image.rfind("sha256:") {
        Some(idx) => {
            let digest = &image[idx + "sha256:".len()..];
            if digest.len() > SHORT_DIGEST {
                format!("{}…", &image[..idx + "sha256:".len() + SHORT_DIGEST])
            } else {
                image.to_string()
            }
        }
        None => image.to_string(),
    }
}

/// Character count, matching Python's `len` on a `str` (not byte length).
fn char_len(s: &str) -> usize {
    s.chars().count()
}

/// Left-pad-free `ljust`: pad a cell on the right to `width` characters.
fn ljust(s: &str, width: usize) -> String {
    let len = char_len(s);
    if len >= width {
        s.to_string()
    } else {
        format!("{s}{}", " ".repeat(width - len))
    }
}

/// Left-aligned columns, sized to their contents, with no trailing space.
pub fn render(table: &[Vec<String>]) -> String {
    if table.is_empty() {
        return String::new();
    }
    let columns = table[0].len();
    let widths: Vec<usize> = (0..columns)
        .map(|c| table.iter().map(|row| char_len(&row[c])).max().unwrap_or(0))
        .collect();
    table
        .iter()
        .map(|row| {
            let joined: Vec<String> = row
                .iter()
                .zip(widths.iter())
                .map(|(cell, w)| ljust(cell, *w))
                .collect();
            joined.join(GAP).trim_end().to_string()
        })
        .collect::<Vec<_>>()
        .join("\n")
}
